File: exercise1/shift_right.py
#!/usr/bin/env python3
#-*- conding: utf-8 -*-
import sys

def reverse(arr, start, end):
	while (start < end):
		temp = arr[start]
		arr[start] = arr[end]
		arr[end] = temp
		start += 1
		end -= 1

def shift_right_by_k(arr, k):
	n = len(arr)
	reverse(arr, 0, n - k - 1)
	reverse(arr, n -k, n - 1)
	reverse(arr, 0, n - 1)

def main():
	arr = [1, 2, 3, 4, 5, 6, 7, 8, 9]
	k = input("please input the number of k: ")
	shift_right_by_k(arr, int(k))
	for a in arr:
		print(a)

if __name__=='__main__':
	main()

File: exercise1/shift_right.c
#include <stdio.h>

/**
 *	将数组array中从start到end的部分（包含start和end位置的元素）反转
 */
void reverse(int *array, int start, int end) {
	int temp;
	while(start < end) {
		temp = array[start];
		array[start++] = array[end];
		array[end--] = temp;
	}
	return;
}

/**
 *	将长度为n的数组array循环右移k位
 */
void shift_right(int *array, int n, int k) {
	reverse(array, 0, n - k - 1);
	reverse(array, n - k, n-1);
	reverse(array, 0, n-1);
}

/**
 *	重置数组中元素的值
 */
void reset(int *a) {
	for (int i = 0; i < 10; i++)
		a[i] = i;
}

int main(int argc, char **argv) {
	int a[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}, k;
	while (1) {
		printf("please the input the number of k(enter 0 to exit): ");
		scanf("%d", &k);
		if (0 == k)
			break;
		k = k % 10;
		shift_right(a, 10, k);
		for (int i = 0; i < 10; i++)
			printf("%d\n", a[i]);
		reset(a);
	}
	return 0;
}
